sphinx_immaterial/python_apigen.py

```diff
@@ -0,0 +1,1511 @@
+        if m is None:
+            break
+        options[m.group(1)] = m.group(2)
+        i += 1
+    assert i < len(rst_strings)
+    assert rst_strings[i] == ""
+
+    i += 1
+    while i < len(rst_strings) and not rst_strings[i].strip():
+        i += 1
+
+    # Strip 3 spaces of indent.
+    content = [line[3:] for line in rst_strings.data[i:]]
+
+    # Extract :group: field if present.
+    group_name: Optional[str] = None
+    group_field_prefix = ":group: "
+    for i, line in enumerate(content):
+        if line.startswith(group_field_prefix):
+            group_name = line[len(group_field_prefix) :].strip()
+            del content[i]
+            break
+
+    return SplitAutodocRstOutput(
+        directive=directive,
+        options=options,
+        content=content,
+        group_name=group_name,
+    )
+
+
+def _summarize_rst_content(content: List[str]) -> List[str]:
+    i = 0
+    # Skip over blank lines before start of directive content
+    while i < len(content) and not content[i].strip():
+        i += 1
+    # Skip over first paragraph
+    while i < len(content) and content[i].strip():
+        i += 1
+
+    return content[:i]
+
+
+class _ApiEntityCollector:
+    def __init__(
+        self,
+        collected_entities: Dict[str, _CollectedApiEntity],
+        strip_module_prefix: str,
+        output_path: str,
+        default_groups: List[Tuple[re.Pattern, str]],
+    ):
+        self.collected_entities = collected_entities
+        self.strip_module_prefix = strip_module_prefix
+        self.output_path = output_path
+        self.default_groups = default_groups
+
+    def collect_entity_recursively(
+        self, entry: _MemberDocumenterEntry, parent_object_name: Optional[str]
+    ) -> None:
+        object_name = entry.object_name
+
+        member_object_names = self.collect_documenter_members(
+            entry.documenter,
+            parent_object_name=object_name,
+        )
+
+        rst_strings = docutils.statemachine.StringList()
+        entry.documenter.directive.result = rst_strings
+        _prepare_documenter_docstring(entry)
+        # Prevent autodoc from also documenting members, since this extension does
+        # that separately.
+        def document_members(*args, **kwargs):
+            return
+
+        entry.documenter.document_members = document_members  # type: ignore[assignment]
+        entry.documenter.generate()
+
+        signatures = entry.documenter.format_signature().split("\n")
+
+        split_result = _split_autodoc_rst_output(rst_strings)
+
+        group_name = split_result.group_name
+        if group_name is None:
+            group_name = _get_group_name(self.default_groups, entry)
+
+        page_name = entry.page_name
+        strip_module_prefix = self.strip_module_prefix
+        if page_name.startswith(strip_module_prefix):
+            page_name = page_name[len(strip_module_prefix) :]
+
+        docname = f"{self.output_path}/{page_name}"
```

Comment:
this isn't cross-platform friendly
```suggestion
        docname = f"{self.output_path}{os.sep}{page_name}"
```
It works fine when pass to functions that have intuition about the platform's path separator (like `open()`), but it doesn't fair well with basic string comparisons.